Significantly improves reading row-based formats with streaming (Kafka, Nuts, RabbitMQ, FileLog)

azat opened the pull request:
### Checklist
- [ ] get info for virtual columns from proper message (right now it get this info from the last message, and tha's why some test fails)

**It will be great if someone will share other feedback, i.e. improvements for other formats (TSV, CSV, JSONEachRow, ...).**

### Changelog category (leave one):
- Performance Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Significantly improves reading row-based formats with streaming (Kafka, Nuts, RabbitMQ, FileLog)

Running format parser from scratch for each messages is too costly,
since formats implemented with batching in mind, and when you parse
single row the overhead is simply too high.

For example this patch **improves performance of ProtobufSingle 7-10x times**
for streaming with Kafka (**from 40K to 390K**).

This patch also changes how error handling works in
StreamingFormatExecutor, before it has separate callback, but now, since
we want to preserve parsing, it uses the same error handling that is
provided by IRowInputFormat, and adjusts
input_format_allow_errors_num=MAX, so that the parser will not stop.

Also now it uses isParseError() in StreamingFormatExecutor, since it
should ignore unknown errors.

Plus, now number of allowed errors is defined on a per-block basis, in
other words before, if you have row-based format, i.e. JSONEachRow, it
is enough to specify kafka_skip_broken_messages=1 to skip any errors,
while now this will not work, as it should (see changes in tests).

And also fix some magic consts in test_kafka_bad_messages test (some
substrings oddities).

azat committed:
Fix significant performance degradation for ProtobufSingle

After #45693 it reset the parser each time, while this is very costly to
do for each row, right now it can handle ~40-50K messages per second
from Kafka engine with ProtobufSingle format, after this change ~70K.

<details>

<summary>perf</summary>

    -   74.72%     1.18%  BgMBSchPool  clickhouse                     [.] DB::StreamingFormatExecutor::execute
       - 73.54% DB::StreamingFormatExecutor::execute
          - 58.14% DB::ISource::work
             - 57.67% DB::ISource::tryGenerate
                - 57.64% DB::IRowInputFormat::generate
                   - 48.65% DB::ProtobufRowInputFormat::readRow
                      - 26.90% DB::ProtobufRowInputFormat::createReaderAndSerializer
                         + 24.95% DB::ProtobufSerializer::create
                         + 0.83% DB::Block::getDataTypes
                         + 0.59% std::__1::vector<std::__1::shared_ptr<DB::IDataType const>, std::__1::allocator<std::__1::shared_ptr<DB::IDataType const> > >::~vector[abi:v15000] (inlined)
                      - 16.94% DB::(anonymous namespace)::ProtobufSerializerMessage::readRow
                         + 6.23% DB::(anonymous namespace)::ProtobufSerializerString<false>::insertDefaults
                         + 5.13% DB::(anonymous namespace)::ProtobufSerializerString<false>::readRow
                         + 2.66% DB::(anonymous namespace)::ProtobufSerializerNumber<unsigned int>::readRow
                         + 1.13% DB::(anonymous namespace)::ProtobufSerializerNumber<unsigned long>::insertDefaults
                           0.56% DB::ProtobufReader::readFieldNumber
                           0.51% DB::(anonymous namespace)::ProtobufSerializerNumber<unsigned long>::readRow
                      - 4.49% DB::(anonymous namespace)::ProtobufSerializerMessage::setColumns
                         + 3.08% DB::(anonymous namespace)::ProtobufSerializerMessage::setColumns
                         + 0.64% COW<DB::IColumn>::getPtr (inlined)
                         + 0.61% std::__1::vector<COW<DB::IColumn>::immutable_ptr<DB::IColumn>, std::__1::allocator<COW<DB::IColumn>::immutable_ptr<DB::IColumn> > >::~vector[abi:v15000] (inlined)
                   + 5.59% DB::Block::cloneEmptyColumns
                   + 2.11% std::__1::vector<COW<DB::IColumn>::mutable_ptr<DB::IColumn>, std::__1::allocator<COW<DB::IColumn>::mutable_ptr<DB::IColumn> > >::~vector[abi:v15000] (inlined)
                     0.51% DB::getDataOffsetMaybeCompressed
          - 12.02% DB::ProtobufRowInputFormat::resetParser
             + 11.89% std::__1::unique_ptr<DB::ProtobufSerializer, std::__1::default_delete<DB::ProtobufSerializer> >::reset[abi:v15000] (inlined)
          + 2.21% DB::ColumnString::insertRangeFrom

</details>

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Significantly improves reading row-based formats with streaming

Running format parser from scratch for each messages is too costly,
since formats implemented with batching in mind, and when you parse
single row the overhead is simply too high.

For example this patch improves performance of ProtobufSingle 7-10x times
for streaming with Kafka (from 40K to 390K).

This patch also changes how error handling works in
StreamingFormatExecutor, before it has separate callback, but now, since
we want to preserve parsing, it uses the same error handling that is
provided by IRowInputFormat, and adjusts
input_format_allow_errors_num=MAX, so that the parser will not stop.

Also now it uses isParseError() in StreamingFormatExecutor, since it
should ignore unknown errors.

Plus, now number of allowed errors is defined on a per-block basis, in
other words before, if you have row-based format, i.e. JSONEachRow, it
is enough to specify kafka_skip_broken_messages=1 to skip any errors,
while now this will not work, as it should (see changes in tests).

And also fix some magic consts in test_kafka_bad_messages test (some
substrings oddities).

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Rework error logging during parsing (add separate interface)

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Adjust number of allowed errors during parsing for StorageFileLog

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Transform avro::Exception to DB::Exception(INCORRECT_DATA)

That way error handling logic in IRowInputFormat will catch it.

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Make ArrowBlockInputFormat retriable (replace UNKNOWN_EXCEPTION with INCORRECT_DATA)

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Make UNKNOWN_TYPE parser error (for i.e. BSONEachRow)

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Mark EMPTY_DATA_PASSED as parser error (for i.e. MySQLDump)

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Mark UNKNOWN_PROTOBUF_FORMAT as parser error

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Mark ATTEMPT_TO_READ_AFTER_EOF as parsing error (for ProtobufSingle)

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Mark IRowInputFormat::read() as final (to avoid possible issues)

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Avoid endless running of test_storage_kafka::test_kafka_settings_new_syntax

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

robot-clickhouse-ci-2 commented:
<!-- automatic status comment for PR #59482 from azat/ClickHouse:row-based-formats-streaming -->
*This is an automated comment for commit 0f106c116aa20a0242cc09d776291d94e96ac56f with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/59482/0f106c116aa20a0242cc09d776291d94e96ac56f/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/59482/0f106c116aa20a0242cc09d776291d94e96ac56f/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

azat committed:
Mark SYNTAX_ERROR as parsing error

Fixes: 02563_async_insert_bad_data
CI: https://s3.amazonaws.com/clickhouse-test-reports/59482/d91a63a43b6e8e487cddca2ec3e6a993a5beef45/fast_test.html
Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Adjust 02889_file_log_save_errors test after error format changes

CI: https://s3.amazonaws.com/clickhouse-test-reports/59482/f1555be976706f2436c2d5b0c64788a583d3899c/fast_test.html
Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Fix ColumnTuple::popBack() when columns contains different number of rows

The problem is that during parsing you can fill only one column of
tuple, but not the other, so you don't need to touch other columns,
since it may leads to bugs, like:

    [localhost] 2024.02.11 19:50:57.660127 [ 9168 ] {e666eadd-ed94-4091-bf2a-a7057d6d102b} <Error> executeQuery: Code: 173. DB::Exception: Amount of memory requested to allocate is more than allowed: (in file/uri /src/ch/clickhouse/.cmake-llvm16/.server/user_files/02475_data.bsonEachRow): While executing BSONEachRowRowInputFormat: While executing File. (CANNOT_ALLOCATE_MEMORY) (version 24.2.1.1) (from [::1]:55432) (comment: 02475_bson_each_row_format.sh) (in query: select * from file(02475_data.bsonEachRow, auto, 'tuple Tuple(x UInt64, b String)')), Stack trace (when copying this message, always include the lines below):

    0. /src/ch/clickhouse/contrib/llvm-project/libcxx/include/exception:141: Poco::Exception::Exception(String const&, int) @ 0x0000000014d1d536 in /src/ch/clickhouse/.cmake-llvm16/programs/clickhouse
    1. /src/ch/clickhouse/src/Common/Exception.cpp:96: DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool) @ 0x000000000b6a30ca in /src/ch/clickhouse/.cmake-llvm16/programs/clickhouse
    2. /src/ch/clickhouse/contrib/llvm-project/libcxx/include/string:1499: DB::Exception::Exception<>(int, FormatStringHelperImpl<>) @ 0x000000000625a017 in /src/ch/clickhouse/.cmake-llvm16/programs/clickhouse
    3. /src/ch/clickhouse/src/Common/PODArray.cpp:29: DB::PODArrayDetails::byte_size(unsigned long, unsigned long) @ 0x000000000b6be4ca in /src/ch/clickhouse/.cmake-llvm16/programs/clickhouse
    4. /src/ch/clickhouse/src/Common/PODArray.h:250: DB::ColumnString::popBack(unsigned long) @ 0x00000000114bddbd in /src/ch/clickhouse/.cmake-llvm16/programs/clickhouse
    5. /src/ch/clickhouse/contrib/llvm-project/libcxx/include/__iterator/wrap_iter.h:100: DB::ColumnTuple::popBack(unsigned long) @ 0x00000000114dd88f in /src/ch/clickhouse/.cmake-llvm16/programs/clickhouse
    6. /src/ch/clickhouse/src/Processors/Formats/IRowInputFormat.cpp:0: DB::IRowInputFormat::read() @ 0x000000001227136f in /src/ch/clickhouse/.cmake-llvm16/programs/clickhouse
    7. /src/ch/clickhouse/src/Processors/Formats/IInputFormat.cpp:19: DB::IInputFormat::generate() @ 0x000000001226d3c7 in /src/ch/clickhouse/.cmake-llvm16/programs/clickhouse
    8. /src/ch/clickhouse/src/Processors/Chunk.h:90: DB::ISource::tryGenerate() @ 0x000000001224b2c7 in /src/ch/clickhouse/.cmake-llvm16/programs/clickhouse
    9. /src/ch/clickhouse/contrib/llvm-project/libcxx/include/optional:344: DB::ISource::work() @ 0x000000001224aed4 in /src/ch/clickhouse/.cmake-llvm16/programs/clickhouse
    10. /src/ch/clickhouse/src/Processors/Executors/ExecutionThreadContext.cpp:0: DB::ExecutionThreadContext::executeTask() @ 0x00000000122638f9 in /src/ch/clickhouse/.cmake-llvm16/programs/clickhouse
    11. /src/ch/clickhouse/src/Processors/Executors/PipelineExecutor.cpp:273: DB::PipelineExecutor::executeStepImpl(unsigned long, std::atomic<bool>*) @ 0x000000001225ab10 in /src/ch/clickhouse/.cmake-llvm16/programs/clickhouse
    12. /src/ch/clickhouse/contrib/llvm-project/libcxx/include/atomic:958: DB::PipelineExecutor::executeStep(std::atomic<bool>*) @ 0x000000001225a41a in /src/ch/clickhouse/.cmake-llvm16/programs/clickhouse
    13. /src/ch/clickhouse/src/Processors/Executors/PullingPipelineExecutor.cpp:54: DB::PullingPipelineExecutor::pull(DB::Chunk&) @ 0x0000000012267a40 in /src/ch/clickhouse/.cmake-llvm16/programs/clickhouse
    14. /src/ch/clickhouse/src/Storages/StorageFile.cpp:1249: DB::StorageFileSource::generate() @ 0x000000001174b5e3 in /src/ch/clickhouse/.cmake-llvm16/programs/clickhouse
    15. /src/ch/clickhouse/src/Processors/Chunk.h:90: DB::ISource::tryGenerate() @ 0x000000001224b2c7 in /src/ch/clickhouse/.cmake-llvm16/programs/clickhouse
    16. /src/ch/clickhouse/contrib/llvm-project/libcxx/include/optional:344: DB::ISource::work() @ 0x000000001224aed4 in /src/ch/clickhouse/.cmake-llvm16/programs/clickhouse
    17. /src/ch/clickhouse/src/Processors/Executors/ExecutionThreadContext.cpp:0: DB::ExecutionThreadContext::executeTask() @ 0x00000000122638f9 in /src/ch/clickhouse/.cmake-llvm16/programs/clickhouse
    18. /src/ch/clickhouse/src/Processors/Executors/PipelineExecutor.cpp:273: DB::PipelineExecutor::executeStepImpl(unsigned long, std::atomic<bool>*) @ 0x000000001225ab10 in /src/ch/clickhouse/.cmake-llvm16/programs/clickhouse
    19. /src/ch/clickhouse/contrib/llvm-project/libcxx/include/__memory/shared_ptr.h:701: DB::PipelineExecutor::executeImpl(unsigned long, bool) @ 0x000000001225a10b in /src/ch/clickhouse/.cmake-llvm16/programs/clickhouse
    20. /src/ch/clickhouse/contrib/llvm-project/libcxx/include/__memory/unique_ptr.h:274: DB::PipelineExecutor::execute(unsigned long, bool) @ 0x0000000012259e53 in /src/ch/clickhouse/.cmake-llvm16/programs/clickhouse
    21. /src/ch/clickhouse/src/Processors/Executors/PullingAsyncPipelineExecutor.cpp:0: void std::__function::__policy_invoker<void ()>::__call_impl<std::__function::__default_alloc_func<ThreadFromGlobalPoolImpl<true>::ThreadFromGlobalPoolImpl<DB::PullingAsyncPipelineExecutor::pull(DB::Chunk&, unsigned long)::$_0>(DB::PullingAsyncPipelineExecutor::pull(DB::Chunk&, unsigned long)::$_0&&)::'lambda'(), void ()>>(std::__function::__policy_storage const*) @ 0x0000000012266cff in /src/ch/clickhouse/.cmake-llvm16/programs/clickhouse
    22. /src/ch/clickhouse/base/base/../base/wide_integer_impl.h:810: ThreadPoolImpl<std::thread>::worker(std::__list_iterator<std::thread, void*>) @ 0x000000000b77ebc6 in /src/ch/clickhouse/.cmake-llvm16/programs/clickhouse
    23. /src/ch/clickhouse/contrib/llvm-project/libcxx/include/__memory/unique_ptr.h:302: void* std::__thread_proxy[abi:v15000]<std::tuple<std::unique_ptr<std::__thread_struct, std::default_delete<std::__thread_struct>>, void ThreadPoolImpl<std::thread>::scheduleImpl<void>(std::function<void ()>, Priority, std::optional<unsigned long>, bool)::'lambda0'()>>(void*) @ 0x000000000b782d62 in /src/ch/clickhouse/.cmake-llvm16/programs/clickhouse
    24. ? @ 0x00007ffff7e3a55a
    25. ? @ 0x00007ffff7eb7a3c

CI: https://s3.amazonaws.com/clickhouse-test-reports/59482/f1555be976706f2436c2d5b0c64788a583d3899c/fast_test.html
Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Fix possible hang during parsing input

The problem is that you cannot check for eof() after parsing in
IRowInputFormat, since this may lead to hang with the following
stacktrace:

    (lldb) bt 20
    * thread 5, name = 'TCPHandler', stop reason = breakpoint 1.1
      * frame 0: 0x0000000016a40d17 clickhouse`__cxa_throw(thrown_object=0x00007ffef5293e80, tinfo=0x0000000001191dd0, dest=(clickhouse`DB::Exception::~Exception() at Exception.h:44)) + 23 at cxa_exception.cpp:258
        frame 1: 0x00000000125804c6 clickhouse`DB::TimeoutReadBufferFromFileDescriptor::nextImpl(this=0x00007ffd619231f0) + 1798 at ShellCommandSource.cpp:0
        frame 2: 0x000000000b716f1e clickhouse`DB::PeekableReadBuffer::nextImpl() [inlined] DB::ReadBuffer::next(this=0x00007ffd619231f0) + 222 at ReadBuffer.h:70
        frame 3: 0x000000000b716f09 clickhouse`DB::PeekableReadBuffer::nextImpl(this=0x00007ffef535e800) + 201 at PeekableReadBuffer.cpp:179
        frame 4: 0x00000000122711ab clickhouse`DB::IRowInputFormat::read() [inlined] DB::ReadBuffer::next(this=0x00007ffef535e800) + 24 at ReadBuffer.h:70
        frame 5: 0x0000000012271193 clickhouse`DB::IRowInputFormat::read() [inlined] DB::ReadBuffer::eof(this=0x00007ffef535e800) + 10 at ReadBuffer.h:112
        frame 6: 0x0000000012271189 clickhouse`DB::IRowInputFormat::read(this=0x00007ffef539ee18) + 873 at IRowInputFormat.cpp:166
        frame 7: 0x000000001226d3c7 clickhouse`DB::IInputFormat::generate(this=<unavailable>) + 39 at IInputFormat.cpp:19
        frame 8: 0x000000001224b2c7 clickhouse`DB::ISource::tryGenerate(this=<unavailable>) + 39 at ISource.cpp:139
        frame 9: 0x000000001224aed4 clickhouse`DB::ISource::work(this=0x00007ffef539ee18) + 52 at ISource.cpp:108
        frame 10: 0x00000000122638f9 clickhouse`DB::ExecutionThreadContext::executeTask() [inlined] DB::executeJob(node=0x00007ffef535eb00, read_progress_callback=0x00007ffd40e0c7e0) + 10 at ExecutionThreadContext.cpp:47
        frame 11: 0x00000000122638ef clickhouse`DB::ExecutionThreadContext::executeTask(this=0x00007ffd61900700) + 239 at ExecutionThreadContext.cpp:95
        frame 12: 0x000000001225ab10 clickhouse`DB::PipelineExecutor::executeStepImpl(this=0x00007ffef536b618, thread_num=0, yield_flag=0x00007ffe728d4dd0) + 816 at PipelineExecutor.cpp:273
        frame 13: 0x000000001225a41a clickhouse`DB::PipelineExecutor::executeStep(this=0x00007ffef536b618, yield_flag=0x00007ffe728d4dd0) + 58 at PipelineExecutor.cpp:148
        frame 14: 0x0000000012267a40 clickhouse`DB::PullingPipelineExecutor::pull(this=0x00007ffe728d4dd0, chunk=0x00007fff007edf70) + 256 at PullingPipelineExecutor.cpp:54
        frame 15: 0x000000001257dd10 clickhouse`DB::(anonymous namespace)::ShellCommandSource::generate(this=0x00007ffd61923000) + 560 at ShellCommandSource.cpp:452

For the test 01746_executable_pool_dictionary.

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Kafka: Obtain information for virtual columns from proper messages

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
KafkaConsumer: Fix initializer for in_use

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat commented:
It is actually too error prone, I need to find a better way.
